namespace/http_resolver.go

```diff
@@ -0,0 +1,423 @@
+			switch tagName, hasAttr := z.TagName(); string(tagName) {
+			case "head":
+				if readingMetaTags {
+					return nil, fmt.Errorf("unexpected head tag")
+				}
+			case "meta":
+				if !hasAttr {
+					return nil, fmt.Errorf("meta tag empty")
+				}
+				scp, newEntries, err := parseHTMLMetaTag(z, name)
+				if err != nil {
+					return nil, err
+				}
+				if scp == "" {
+					for _, entry := range newEntries {
+						if err = entries.Add(entry); err != nil {
+							return nil, err
+						}
+					}
+				} else if parsedScope != "" {
+					return nil, fmt.Errorf("multiple scopes defined")
+				} else {
+					parsedScope = scp
+				}
+			default:
+				return nil, fmt.Errorf("unexpected html element %q", string(tagName))
+			}
+			readingMetaTags = true
+		case html.EndTagToken:
+			switch tagName, _ := z.TagName(); string(tagName) {
+			case "head":
+				break ParsingLoop
+			case "meta":
+			default:
+				return nil, fmt.Errorf("unexpected tag %q", tagName)
+			}
+		default:
+			continue ParsingLoop
+		}
+	}
+	if !readingMetaTags || len(entries.entries) == 0 {
+		return nil, fmt.Errorf("no entries found")
+	}
+	if parsedScope == "" {
+		parsedScope, err = parseScope(name)
+		if err != nil {
+			return nil, fmt.Errorf("cannot use given name %q as scope: %v", name, err)
+		}
+	}
+	for i := range entries.entries {
+		entries.entries[i].scope = parsedScope
+	}
+	return entries, nil
+}
+
+type httpResolver struct {
+	config *HTTPResolverConfig
+	cache  map[string]cacheEntry
+}
+
+/* Create base HTTP resolver.
+ *
+ * It uses discovery process to fetch scope entries. Multiple discovery
+ * endpoints may be queried during single `Resolve()` call depending on scope's
+ * namespace extensions and given callback.
+ *
+ * It caches fetched entries for further lookups.
+ */
+func NewHTTPResolver(config *HTTPResolverConfig) Resolver {
+	if config == nil {
+		config = &HTTPResolverConfig{}
+	}
+	if config.Client == nil {
+		config.Client = &http.Client{}
+	}
+	if config.ResolverFactory == nil {
+		config.ResolverFactory = func(entries *Entries) Resolver {
+			return NewSimpleResolver(entries, true)
+		}
+	}
+	if config.NSResolveCallback == nil {
+		config.NSResolveCallback = func(name string, namespace scope) NSResolveActionEnum {
+			if !namespace.Contains(name) {
+				logrus.Debugf("Ignoring extension namespace %q which isn't an ancestor of %q", namespace, name)
+				return NSResolveActionIgnore
+			}
+			return NSResolveActionRecurse
+		}
+	}
+	return &httpResolver{
+		config: config,
+		cache:  make(map[string]cacheEntry),
+	}
+}
+
+func (hr *httpResolver) nameToURL(name string) string {
+	return "https://" + name + "?docker-discovery=1"
+}
+
+func (hr *httpResolver) resolveEntries(es *Entries, visited map[string]struct{}, name string) error {
+	cached, exists := hr.cache[name]
+	if exists && !cached.created.Add(hr.config.ExpireAfter).Before(time.Now()) {
+		entries, err := hr.config.ResolverFactory(cached.entries).Resolve(name)
```

Comment:
Yeah, kind of. It has a confusing name for that purpose though. Now I think it would be better to use the same resolver both for cached and currently fetched entries.

The thing is I'd like to use `simpleResolver`'s resolve algorithm on fetched/cached entries rather than duplicate it here. Neither `multiResolver` nor `extendResolver` provide a way to pass entries obtained from one resolver to another and run `Resolve()` on them. Maybe some sort of:

    type ResolverFactory func(*Entries) Resolver
    type recursiveResolver struct {
        baseResolver Resolver
        resolverFactories []ResolverFactory
    }

    func (cr *recursiveResolver) Resolve(name string) (*Entries, error) {
        entries, err := cr.baseResolver.Resolver(name)
        if err != nil {
            return nil, err
        }
        for _, f := cr.resolverFactories {
            entries, err := f(entries).Resolve(name)
            if err != nil {
                return nil, err
            }
        }
        return entries, err
    }

would do the trick. So I could write:

    resolver := NewRecursiveResolver(NewHTTPResolver(...), func(entries *Entries) Resolver { return NewSimpleResolver(entries) })

WDYT?